src/hardware/vga_memory.cpp

```diff
@@ -1057,24 +1058,35 @@ void VGA_StartUpdateLFB(void) {
 static void VGA_Memory_ShutDown(Section * /*sec*/) {
 #ifdef VGA_KEEP_CHANGES
 	delete[] vga.changes.map;
+	vga.mem.linear = {};
+	vga.fastmem    = {};
 #endif
 }
 
-void VGA_SetupMemory(Section* sec) {
+// The video memory is read from and written to in sizes up to uint32_t
+using video_chunk_t = uint32_t;
+static std::vector<video_chunk_t> linear_buffer  = {};
+static std::vector<video_chunk_t> fastmem_buffer = {};
+
+void VGA_SetupMemory(Section* sec)
+{
 	vga.svga.bank_read = vga.svga.bank_write = 0;
 	vga.svga.bank_read_full = vga.svga.bank_write_full = 0;
 
-	// ensure memory is aligned to vga_memalign bytes
-	assert(reinterpret_cast<uintptr_t>(vga.mem.linear) % vga_memalign == 0);
-	assert(reinterpret_cast<uintptr_t>(vga.fastmem) % vga_memalign == 0);
+	// lower limit at 512k plus an 2 KB for one scan-line
+	constexpr uint32_t vga_mem_bytes_min        = 512 * 1024;
+	constexpr uint32_t vga_mem_scanline_reserve = 2048;
 
-	uint32_t vga_allocsize=vga.vmemsize;
-	// Keep lower limit at 512k
-	if (vga_allocsize<512*1024) vga_allocsize=512*1024;
-	// We reserve extra 2K for one scan line
-	vga_allocsize+=2048;
+	// Allocate and assign the linear frame-buffer
+	const auto num_linear_bytes = std::max(vga_mem_bytes_min, vga.vmemsize) +
+	                              vga_mem_scanline_reserve;
+	linear_buffer.resize(num_linear_bytes / sizeof(video_chunk_t));
+	vga.mem.linear = reinterpret_cast<uint8_t*>(linear_buffer.data());
 
-	memset(vga.mem.linear,0,vga_allocsize);
+	// Allocate and assign the fast-memory buffer
+	const auto num_fastmem_bytes = 2 * num_linear_bytes;
+	fastmem_buffer.resize(num_fastmem_bytes / sizeof(video_chunk_t));
+	vga.fastmem = reinterpret_cast<uint8_t*>(fastmem_buffer.data());
```

Comment:
Previous code used `alignas(vga_memalign)` specifier for both `fastmem` and `linear`. Is there anything that guarantees proper memory alignment after the refactoring? Similarly with `MemBase`.

Reply:
From what I can see, the largest type handled by the video array is `uint32_t`, so we need to at least have alignment against that.

However, there is mention of operations done in 16-byte chunks, but I can't see where it happens on the host side - so I'll assume it's done on the DOS side. So I've made the vector's type a 16-byte struct.

All `new` T and container<T> allocations are aligned on the type's size; and I added an assert to confirm that's the case for both of the video mem vectors:

``` c++
assert(reinterpret_cast<uintptr_t>(vga.fastmem) % video_chunk_bytes == 0);
```

Reply:
to add a little to the motivation behind aligning to the host pagesize, looking at a couple of functions that did calculate offsets in multiples of dospagesize, we already needed to align it anyways to e.g. uint64_t, I just went with hostpage size.

```cpp
	HostPt GetHostReadPt(Bitu phys_page) {
		return MemBase+phys_page*dos_pagesize;
	}
	HostPt GetHostWritePt(Bitu phys_page) {
		return MemBase+phys_page*dos_pagesize;
	}
```

Reply:
@kcgen Aligning CPU-visible memory to the host page size will be definitely needed if we want to support virtualization instead of CPU emulation in the future.

Reply:
Updated with commit d66b3fe7becf8cb5f6b253b507b13a5c4dfb16a7, which now uses actual 4096-byte pages in the memory.pages vector, so instead of this:

```c++
return MemBase+phys_page*dos_pagesize;
```

We can now access pages themselves, and even bytes within the pages (and its all contiguous too):

```c++
return &(memory.pages[phys_page].bytes[0]);
```

This type of access makes the code less opaque as to what's going on, so this page-construction is definitely a win. Thanks @kklobe and @FeralChild64.